transport/http/outbound.go

```diff
@@ -41,64 +45,153 @@ var (
+	Path     string
+	Scheme   string
 }
 
 func (o *outbound) Start(d transport.Deps) error {
 	if o.started.Swap(true) {
 		return errOutboundAlreadyStarted
 	}
 	o.Deps = d
-	return nil
+	return o.PeerList.Start()
 }
 
 func (o *outbound) Stop() error {
 	if !o.started.Swap(false) {
 		return errOutboundNotStarted
 	}
-	return nil
+	return o.PeerList.Stop()
 }
 
-func (o *outbound) createSpan(ctx context.Context, req *http.Request, treq *transport.Request, start time.Time) (context.Context, opentracing.Span) {
+func (o *outbound) Call(ctx context.Context, treq *transport.Request) (*transport.Response, error) {
+	if !o.started.Load() {
+		// panic because there's no recovery from this
+		panic(errOutboundNotStarted)
+	}
+	start := time.Now()
+	deadline, _ := ctx.Deadline()
+	ttl := deadline.Sub(start)
+
+	peer, err := o.getPeerForRequest(ctx, treq)
+	if err != nil {
+		return nil, err
+	}
+	endRequest := peer.StartRequest()
+	defer endRequest()
+
+	req, err := o.createRequest(peer, treq)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header = applicationHeaders.ToHTTPHeaders(treq.Headers, nil)
+	ctx, req, span := o.withOpentracingSpan(ctx, req, treq, start)
+	defer span.Finish()
+	req = o.withCoreHeaders(req, treq, ttl)
+
+	client, err := o.getHTTPClient(peer)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := client.Do(req.WithContext(ctx))
+
+	if err != nil {
+		// Workaround borrowed from ctxhttp until
+		// https://github.com/golang/go/issues/17711 is resolved.
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		default:
+		}
+
+		span.SetTag("error", true)
+		span.LogEvent(err.Error())
+		if err == context.DeadlineExceeded {
+			end := time.Now()
+			return nil, errors.ClientTimeoutError(treq.Service, treq.Procedure, end.Sub(start))
+		}
+
+		return nil, err
+	}
+
+	span.SetTag("http.status_code", response.StatusCode)
+
+	if response.StatusCode >= 200 && response.StatusCode < 300 {
+		appHeaders := applicationHeaders.FromHTTPHeaders(
+			response.Header, transport.NewHeaders())
+		return &transport.Response{
+			Headers: appHeaders,
+			Body:    response.Body,
+		}, nil
+	}
+
+	return nil, getErrFromResponse(response)
+}
+
+func (o *outbound) getPeerForRequest(ctx context.Context, treq *transport.Request) (*hostport.Peer, error) {
+	peer, err := o.PeerList.ChoosePeer(ctx, treq)
+	if err != nil {
+		return nil, err
+	}
+
+	hpPeer, ok := peer.(*hostport.Peer)
+	if ok {
+		return hpPeer, nil
+
+	}
+
+	return nil, terrors.ErrInvalidPeerConversion{
+		Peer:         peer,
+		ExpectedType: "*hostport.Peer",
+	}
+}
+
+func (o *outbound) createRequest(peer *hostport.Peer, treq *transport.Request) (*http.Request, error) {
+	reqURL := fmt.Sprintf("%s://%s%s", o.Scheme, peer.HostPort(), o.Path)
+	return http.NewRequest("POST", reqURL, treq.Body)
+}
+
+func (o *outbound) withOpentracingSpan(ctx context.Context, req *http.Request, treq *transport.Request, start time.Time) (context.Context, *http.Request, opentracing.Span) {
```

Comment:
why the rename (from createSpan)? outbound.withOpentracingSpan(..) sounds like the span is being augmented by a span, when really we're creating it from the context/request object

also why not drop 'Opentracing' (withSpan), seems redundant naming the package name